File: javascript/test/resources.test.ts
/** Every resource method targets the URL and body the API documents. */

import { describe, expect, it } from 'vitest';

import { KYCCentral } from '../src/index.js';
import { BASE_URL, jsonResponse, mockFetch } from './helpers.js';

function client(): { client: KYCCentral; fetch: ReturnType<typeof mockFetch> } {
  const fetch = mockFetch([jsonResponse({})]);
  return { client: new KYCCentral({ apiKey: 'k', baseUrl: BASE_URL, fetch }), fetch };
}

const CASES: Array<[string, (c: KYCCentral) => Promise<unknown>, string]> = [
  ['companies.get', (c) => c.companies.get('00445790'), '/v1/companies/00445790'],
  ['companies.dossier', (c) => c.companies.dossier('00445790'), '/v1/companies/00445790/dossier'],
  [
    'companies.officers',
    (c) => c.companies.officers('00445790'),
    '/v1/companies/00445790/officers',
  ],
  [
    'companies.pscs',
    (c) => c.companies.pscs('00445790'),
    '/v1/companies/00445790/persons-with-significant-control',
  ],
  [
    'companies.pscStatements',
    (c) => c.companies.pscStatements('00445790'),
    '/v1/companies/00445790/persons-with-significant-control-statements',
  ],
  [
    'companies.pscChainDepth',
    (c) => c.companies.pscChainDepth('00445790'),
    '/v1/companies/00445790/psc-chain-depth',
  ],
  [
    'companies.pscChainTree',
    (c) => c.companies.pscChainTree('00445790'),
    '/v1/companies/00445790/psc-chain-tree',
  ],
  ['companies.charges', (c) => c.companies.charges('00445790'), '/v1/companies/00445790/charges'],
  [
    'companies.charge',
    (c) => c.companies.charge('00445790', 'chg1'),
    '/v1/companies/00445790/charges/chg1',
  ],
  [
    'companies.insolvency',
    (c) => c.companies.insolvency('00445790'),
    '/v1/companies/00445790/insolvency',
  ],
  [
    'companies.disqualifications',
    (c) => c.companies.disqualifications('00445790'),
    '/v1/companies/00445790/disqualifications',
  ],
  [
    'companies.officerDisqualification',
    (c) => c.companies.officerDisqualification('00445790', 'off1'),
    '/v1/companies/00445790/officers/off1/disqualification',
  ],
  [
    'companies.filingHistory',
    (c) => c.companies.filingHistory('00445790'),
    '/v1/companies/00445790/filing-history',
  ],
  [
    'companies.filingExtract',
    (c) => c.companies.filingExtract('00445790', 'tx1'),
    '/v1/companies/00445790/filing-history/tx1/extract',
  ],
  [
    'companies.statementOfCapital',
    (c) => c.companies.statementOfCapital('00445790'),
    '/v1/companies/00445790/statement-of-capital',
  ],
  ['ruleSets.list', (c) => c.ruleSets.list(), '/v1/rule-sets'],
  ['rules.list', (c) => c.rules.list(), '/v1/rules'],
  ['rules.fields', (c) => c.rules.fields(), '/v1/rules/fields'],
  ['jobs.get', (c) => c.jobs.get('job-1'), '/v1/jobs/job-1'],
  ['jurisdictions.list', (c) => c.jurisdictions.list(), '/v1/jurisdictions'],
  ['jurisdictions.check', (c) => c.jurisdictions.check('Iran'), '/v1/jurisdictions/check'],
  [
    'offshoreJurisdictions.list',
    (c) => c.offshoreJurisdictions.list(),
    '/v1/offshore-jurisdictions',
  ],
  ['sanctions.status', (c) => c.sanctions.status(), '/v1/sanctions/status'],
  ['sanctions.meta', (c) => c.sanctions.meta(), '/v1/sanctions/meta'],
  ['sanctions.screen', (c) => c.sanctions.screen('Alice'), '/v1/sanctions/screen'],
  ['sanctions.entities', (c) => c.sanctions.entities(), '/v1/sanctions/entities'],
  ['news.status', (c) => c.news.status(), '/v1/news/status'],
  ['news.screenCompany', (c) => c.news.screenCompany('00445790'), '/v1/news/screen-company'],
  ['offshoreLeaks.status', (c) => c.offshoreLeaks.status(), '/v1/offshore-leaks/status'],
  ['offshoreLeaks.node', (c) => c.offshoreLeaks.node('n1'), '/v1/offshore-leaks/node/n1'],
  [
    'offshoreLeaks.screenCompany',
    (c) => c.offshoreLeaks.screenCompany('00445790'),
    '/v1/offshore-leaks/screen-company',
  ],
  ['fca.status', (c) => c.fca.status(), '/v1/fca/status'],
  ['fca.search', (c) => c.fca.search('acme'), '/v1/fca/search'],
  ['fca.firm', (c) => c.fca.firm('123456'), '/v1/fca/firm/123456'],
  ['fca.firmNames', (c) => c.fca.firmNames('123456'), '/v1/fca/firm/123456/names'],
  [
    'fca.firmIndividuals',
    (c) => c.fca.firmIndividuals('123456'),
    '/v1/fca/firm/123456/individuals',
  ],
  [
    'fca.screenIndividuals',
    (c) => c.fca.screenIndividuals('00445790'),
    '/v1/fca/screen-individuals',
  ],
  ['fca.checkIndividual', (c) => c.fca.checkIndividual('Alice'), '/v1/fca/check-individual'],
  ['gleif.company', (c) => c.gleif.company('00445790'), '/v1/gleif/company'],
  [
    'individualInsolvency.screenCompany',
    (c) => c.individualInsolvency.screenCompany('00445790'),
    '/v1/individual-insolvency/screen-company',
  ],
  ['charity.status', (c) => c.charity.status(), '/v1/charity/status'],
  ['charity.search', (c) => c.charity.search('oxfam'), '/v1/charity/search'],
  ['charity.get', (c) => c.charity.get('1234'), '/v1/charity/charity/1234'],
  ['charity.trustees', (c) => c.charity.trustees('1234'), '/v1/charity/charity/1234/trustees'],
  ['hmrcVat.status', (c) => c.hmrcVat.status(), '/v1/hmrc-vat/status'],
  ['hmrcVat.check', (c) => c.hmrcVat.check('GB123456789'), '/v1/hmrc-vat/check'],
  ['analysis.status', (c) => c.analysis.status(), '/v1/analysis/status'],
  ['dataSourceHealth', (c) => c.dataSourceHealth(), '/health/data-sources'],
];

describe('endpoint routing', () => {
  it.each(CASES)('%s hits %s', async (_name, call, path) => {
    const { client: c, fetch } = client();
    await call(c);
    expect(new URL(fetch.mock.calls[0]![0]).pathname).toBe(path);
  });
});

describe('request bodies', () => {
  it('posts a names body for batch sanctions screening', async () => {
    const { client: c, fetch } = client();
    await c.sanctions.screenNames(['Alice Example', '  Bob Example  ', '']);

    const body = JSON.parse(fetch.mock.calls[0]![1]!.body as string);
    // Blank entries dropped, surrounding whitespace trimmed.
    expect(body).toEqual({ names: ['Alice Example', 'Bob Example'] });
  });

  it('rejects empty name batches', async () => {
    const { client: c } = client();
    await expect(c.sanctions.screenNames([])).rejects.toThrow(/at least one/);
    await expect(c.offshoreLeaks.screenNames(['', '  '])).rejects.toThrow(/at least one/);
  });

  it('enforces the documented batch cap', async () => {
    const { client: c } = client();
    const names = Array.from({ length: 501 }, (_, i) => `name ${i}`);
    await expect(c.sanctions.screenNames(names)).rejects.toThrow(/at most 500/);
  });

  it('validates the entity search shape', async () => {
    const { client: c, fetch } = client();
    await c.news.searchEntities([{ key: 'psc-1', names: ['Alice Example', 'A. Example'] }]);

    const body = JSON.parse(fetch.mock.calls[0]![1]!.body as string);
    expect(body).toEqual({ entities: [{ key: 'psc-1', names: ['Alice Example', 'A. Example'] }] });

    await expect(c.news.searchEntities([{ key: '', names: ['Alice'] }])).rejects.toThrow(/'key'/);
    await expect(c.news.searchEntities([{ key: 'psc-1', names: [] }])).rejects.toThrow(
      /at least one name/,
    );
  });

  it('builds the analysis body', async () => {
    const { client: c, fetch } = client();
    await c.analysis.company('00445790', { ruleSetId: 'quick', sections: ['ownership'] });

    const body = JSON.parse(fetch.mock.calls[0]![1]!.body as string);
    expect(body).toEqual({
      company_number: '00445790',
      rule_set_id: 'quick',
      sections: ['ownership'],
    });
  });

  it('validates the filing extract mode', async () => {
    const { client: c } = client();
    await expect(
      c.analysis.filingExtract('00445790', 'tx1', { mode: 'summarise' as never }),
    ).rejects.toThrow(/mode must be one of/);
  });

  it('validates documentation questions', async () => {
    const { client: c } = client();
    await expect(c.docs.ask('   ')).rejects.toThrow(/non-empty/);
    await expect(
      c.docs.ask('hi', { history: Array(13).fill({ role: 'user', content: 'x' }) }),
    ).rejects.toThrow(/at most 12/);
  });

  it('repeats list filters in advanced search', async () => {
    const { client: c, fetch } = client();
    await c.companies.advancedSearch({
      companyNameIncludes: 'tesco',
      companyStatus: ['active', 'liquidation'],
    });

    const url = new URL(fetch.mock.calls[0]![0]);
    expect(url.searchParams.getAll('company_status')).toEqual(['active', 'liquidation']);
    expect(url.searchParams.get('company_name_includes')).toBe('tesco');
  });
});

File: elixir/test/kyccentral/resources_test.exs
defmodule KYCCentral.ResourcesTest do
  @moduledoc "Every resource function targets the URL and body the API documents."

  use ExUnit.Case, async: true

  alias KYCCentral.{Error, Stub}

  # Stored as module/function/extra-argument tuples rather than anonymous
  # functions: the list is escaped into generated tests at compile time, and
  # closures cannot be escaped.
  @routes [
    {KYCCentral.Companies, :get, ["00445790"], "/v1/companies/00445790"},
    {KYCCentral.Companies, :dossier, ["00445790"], "/v1/companies/00445790/dossier"},
    {KYCCentral.Companies, :officers, ["00445790"], "/v1/companies/00445790/officers"},
    {KYCCentral.Companies, :pscs, ["00445790"],
     "/v1/companies/00445790/persons-with-significant-control"},
    {KYCCentral.Companies, :psc_statements, ["00445790"],
     "/v1/companies/00445790/persons-with-significant-control-statements"},
    {KYCCentral.Companies, :psc_chain_depth, ["00445790"],
     "/v1/companies/00445790/psc-chain-depth"},
    {KYCCentral.Companies, :psc_chain_tree, ["00445790"],
     "/v1/companies/00445790/psc-chain-tree"},
    {KYCCentral.Companies, :charges, ["00445790"], "/v1/companies/00445790/charges"},
    {KYCCentral.Companies, :charge, ["00445790", "chg1"], "/v1/companies/00445790/charges/chg1"},
    {KYCCentral.Companies, :insolvency, ["00445790"], "/v1/companies/00445790/insolvency"},
    {KYCCentral.Companies, :disqualifications, ["00445790"],
     "/v1/companies/00445790/disqualifications"},
    {KYCCentral.Companies, :officer_disqualification, ["00445790", "off1"],
     "/v1/companies/00445790/officers/off1/disqualification"},
    {KYCCentral.Companies, :officer_appointments, ["off1"],
     "/v1/companies/officers/off1/appointments"},
    {KYCCentral.Companies, :filing_history, ["00445790"],
     "/v1/companies/00445790/filing-history"},
    {KYCCentral.Companies, :filing_extract, ["00445790", "tx1"],
     "/v1/companies/00445790/filing-history/tx1/extract"},
    {KYCCentral.Companies, :statement_of_capital, ["00445790"],
     "/v1/companies/00445790/statement-of-capital"},
    {KYCCentral.Companies, :search, ["tesco"], "/v1/companies/search"},
    {KYCCentral.Companies, :search_officers, ["smith"], "/v1/companies/search/officers"},
    {KYCCentral.Companies, :advanced_search, [], "/v1/companies/advanced-search"},
    {KYCCentral.RuleSets, :list, [], "/v1/rule-sets"},
    {KYCCentral.Rules, :list, [], "/v1/rules"},
    {KYCCentral.Rules, :fields, [], "/v1/rules/fields"},
    {KYCCentral.Jobs, :get, ["job-1"], "/v1/jobs/job-1"},
    {KYCCentral.Jurisdictions, :list, [], "/v1/jurisdictions"},
    {KYCCentral.Jurisdictions, :check, ["Iran"], "/v1/jurisdictions/check"},
    {KYCCentral.OffshoreJurisdictions, :list, [], "/v1/offshore-jurisdictions"},
    {KYCCentral.OffshoreJurisdictions, :check, ["Panama"], "/v1/offshore-jurisdictions/check"},
    {KYCCentral.Sanctions, :status, [], "/v1/sanctions/status"},
    {KYCCentral.Sanctions, :meta, [], "/v1/sanctions/meta"},
    {KYCCentral.Sanctions, :screen, ["Alice"], "/v1/sanctions/screen"},
    {KYCCentral.Sanctions, :entities, [], "/v1/sanctions/entities"},
    {KYCCentral.News, :status, [], "/v1/news/status"},
    {KYCCentral.News, :screen_company, ["00445790"], "/v1/news/screen-company"},
    {KYCCentral.OffshoreLeaks, :status, [], "/v1/offshore-leaks/status"},
    {KYCCentral.OffshoreLeaks, :node, ["n1"], "/v1/offshore-leaks/node/n1"},
    {KYCCentral.OffshoreLeaks, :screen_company, ["00445790"],
     "/v1/offshore-leaks/screen-company"},
    {KYCCentral.FCA, :status, [], "/v1/fca/status"},
    {KYCCentral.FCA, :search, ["acme"], "/v1/fca/search"},
    {KYCCentral.FCA, :firm, ["123456"], "/v1/fca/firm/123456"},
    {KYCCentral.FCA, :firm_names, ["123456"], "/v1/fca/firm/123456/names"},
    {KYCCentral.FCA, :firm_individuals, ["123456"], "/v1/fca/firm/123456/individuals"},
    {KYCCentral.FCA, :screen_individuals, ["00445790"], "/v1/fca/screen-individuals"},
    {KYCCentral.FCA, :check_individual, ["Alice"], "/v1/fca/check-individual"},
    {KYCCentral.GLEIF, :company, ["00445790"], "/v1/gleif/company"},
    {KYCCentral.IndividualInsolvency, :screen_company, ["00445790"],
     "/v1/individual-insolvency/screen-company"},
    {KYCCentral.Charity, :status, [], "/v1/charity/status"},
    {KYCCentral.Charity, :search, ["oxfam"], "/v1/charity/search"},
    {KYCCentral.Charity, :get, ["1234"], "/v1/charity/charity/1234"},
    {KYCCentral.Charity, :trustees, ["1234"], "/v1/charity/charity/1234/trustees"},
    {KYCCentral.HMRCVat, :status, [], "/v1/hmrc-vat/status"},
    {KYCCentral.HMRCVat, :check, ["GB123456789"], "/v1/hmrc-vat/check"},
    {KYCCentral.Analysis, :status, [], "/v1/analysis/status"},
    {KYCCentral, :data_source_health, [], "/health/data-sources"}
  ]

  for {module, function, args, path} <- @routes do
    test "#{inspect(module)}.#{function} hits #{path}" do
      {client, agent} = Stub.client([Stub.json(%{})])

      assert {:ok, _} = apply(unquote(module), unquote(function), [client | unquote(args)])
      assert Stub.path(agent) == unquote(path)
    end
  end

  describe "batch screening bodies" do
    test "posts a names body" do
      {client, agent} = Stub.client([Stub.json(%{})])

      assert {:ok, _} =
               KYCCentral.Sanctions.screen_names(client, ["Alice Example", "  Bob Example  ", ""])

      # Blank entries dropped, surrounding whitespace trimmed.
      assert Stub.body(agent) == %{"names" => ["Alice Example", "Bob Example"]}
      assert Stub.path(agent) == "/v1/sanctions/screen-names"
    end

    test "rejects an empty batch before making a request" do
      {client, agent} = Stub.client([Stub.json(%{})])

      assert {:error, %Error{kind: :invalid_argument}} =
               KYCCentral.Sanctions.screen_names(client, [])

      assert {:error, %Error{kind: :invalid_argument}} =
               KYCCentral.OffshoreLeaks.screen_names(client, ["", "   "])

      assert Stub.call_count(agent) == 0
    end

    test "enforces the documented batch cap" do
      {client, _agent} = Stub.client([Stub.json(%{})])
      names = Enum.map(1..501, &"name #{&1}")

      assert {:error, %Error{kind: :invalid_argument, message: message}} =
               KYCCentral.Sanctions.screen_names(client, names)

      assert message =~ "at most 500"
    end

    test "adverse media has no client-side cap, since the API's is tunable" do
      {client, agent} = Stub.client([Stub.json(%{})])
      names = Enum.map(1..501, &"name #{&1}")

      assert {:ok, _} = KYCCentral.News.search_names(client, names)
      assert length(Stub.body(agent)["names"]) == 501
    end
  end

  describe "entity search" do
    test "accepts atom- and string-keyed entities" do
      {client, agent} = Stub.client([Stub.json(%{})])

      assert {:ok, _} =
               KYCCentral.News.search_entities(client, [
                 %{key: "psc-1", names: ["Alice Example", "A. Example"]},
                 %{"key" => "psc-2", "names" => ["Bob Example"]}
               ])

      assert Stub.body(agent) == %{
               "entities" => [
                 %{"key" => "psc-1", "names" => ["Alice Example", "A. Example"]},
                 %{"key" => "psc-2", "names" => ["Bob Example"]}
               ]
             }
    end

    test "validates the entity shape" do
      {client, _agent} = Stub.client([Stub.json(%{})])

      assert {:error, %Error{kind: :invalid_argument, message: key_message}} =
               KYCCentral.News.search_entities(client, [%{names: ["Alice"]}])

      assert key_message =~ ":key"

      assert {:error, %Error{kind: :invalid_argument, message: names_message}} =
               KYCCentral.News.search_entities(client, [%{key: "psc-1", names: []}])

      assert names_message =~ "at least one name"

      assert {:error, %Error{kind: :invalid_argument}} =
               KYCCentral.News.search_entities(client, [])
    end
  end

  describe "analysis and docs" do
    test "builds the analysis body" do
      {client, agent} = Stub.client([Stub.json(%{})])

      assert {:ok, _} =
               KYCCentral.Analysis.company(client, "00445790",
                 rule_set_id: "quick",
                 sections: ["ownership"]
               )

      assert Stub.body(agent) == %{
               "company_number" => "00445790",
               "rule_set_id" => "quick",
               "sections" => ["ownership"]
             }
    end

    test "validates the filing extract mode" do
      {client, _agent} = Stub.client([Stub.json(%{})])

      assert {:error, %Error{kind: :invalid_argument, message: message}} =
               KYCCentral.Analysis.filing_extract(client, "00445790", "tx1", mode: "summarise")

      assert message =~ "must be one of"
    end

    test "defaults the filing extract mode to extract" do
      {client, agent} = Stub.client([Stub.json(%{})])

      assert {:ok, _} = KYCCentral.Analysis.filing_extract(client, "00445790", "tx1")
      assert Stub.body(agent)["mode"] == "extract"
    end

    test "validates documentation questions" do
      {client, _agent} = Stub.client([Stub.json(%{})])

      assert {:error, %Error{kind: :invalid_argument}} = KYCCentral.Docs.ask(client, "   ")

      history = List.duplicate(%{role: "user", content: "x"}, 13)

      assert {:error, %Error{kind: :invalid_argument, message: message}} =
               KYCCentral.Docs.ask(client, "hi", history: history)

      assert message =~ "at most 12"
    end
  end

  test "advanced search repeats list filters" do
    {client, agent} = Stub.client([Stub.json(%{})])

    assert {:ok, _} =
             KYCCentral.Companies.advanced_search(client,
               company_name_includes: "tesco",
               company_status: ["active", "liquidation"]
             )

    assert Stub.query_values(agent, "company_status") == ["active", "liquidation"]
    assert Stub.query_value(agent, "company_name_includes") == "tesco"
  end
end
